types/wlr_foreign_toplevel_v1.c

```diff
@@ -0,0 +1,477 @@
+	wl_signal_init(&toplevel->events.request_close);
+
+	struct wl_resource *manager_resource, *tmp;
+	wl_resource_for_each_safe(manager_resource, tmp, &manager->resources) {
+		create_toplevel_resource_for_resource(toplevel, manager_resource);
+	}
+
+	return toplevel;
+}
+
+static const struct zwlr_foreign_toplevel_manager_v1_interface
+	foreign_toplevel_manager_impl;
+
+static void foreign_toplevel_manager_handle_stop(struct wl_client *client,
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource,
+		&zwlr_foreign_toplevel_manager_v1_interface,
+		&foreign_toplevel_manager_impl));
+
+	zwlr_foreign_toplevel_manager_v1_send_finished(resource);
+	wl_resource_destroy(resource);
+}
+
+static const struct zwlr_foreign_toplevel_manager_v1_interface
+		foreign_toplevel_manager_impl = {
+	.stop = foreign_toplevel_manager_handle_stop
+};
+
+static void foreign_toplevel_manager_resource_destroy(
+		struct wl_resource *resource) {
+	wl_list_remove(wl_resource_get_link(resource));
+}
+
+static void toplevel_send_details_to_toplevel_resource(
+		struct wlr_foreign_toplevel_v1 *toplevel, struct wl_resource *resource) {
+	if (toplevel->title) {
+		zwlr_foreign_toplevel_v1_send_title(resource, toplevel->title);
+	}
+	if (toplevel->app_id) {
+		zwlr_foreign_toplevel_v1_send_app_id(resource, toplevel->app_id);
+	}
+
+	struct wlr_foreign_toplevel_output *output;
+	wl_list_for_each(output, &toplevel->outputs, link) {
+		send_output_to_resource(resource, output->output, true);
+	}
+
+	struct wl_array states;
+	wl_array_init(&states);
+	fill_array_from_toplevel_state(&states, toplevel->state);
+	zwlr_foreign_toplevel_v1_send_state(resource, &states);
+	wl_array_release(&states);
+
+	zwlr_foreign_toplevel_v1_send_done(resource);
+}
+
+static void foreign_toplevel_manager_bind(struct wl_client *client, void *data,
+		uint32_t version, uint32_t id) {
+	struct wlr_foreign_toplevel_manager_v1 *manager = data;
+	assert(client && manager);
+
+	struct wl_resource *resource = wl_resource_create(client,
+			&zwlr_foreign_toplevel_manager_v1_interface, version, id);
+	if (!resource) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+	wl_resource_set_implementation(resource, &foreign_toplevel_manager_impl,
+			manager, foreign_toplevel_manager_resource_destroy);
+
+	wl_list_insert(&manager->resources, wl_resource_get_link(resource));
+
+	struct wlr_foreign_toplevel_v1 *toplevel, *tmp;
+	wl_list_for_each_safe(toplevel, tmp, &manager->toplevels, link) {
+		struct wl_resource *toplevel_resource =
+			create_toplevel_resource_for_resource(toplevel, resource);
+		toplevel_send_details_to_toplevel_resource(toplevel, toplevel_resource);
+	}
+}
+
+void wlr_foreign_toplevel_manager_v1_destroy(
+		struct wlr_foreign_toplevel_manager_v1 *manager) {
+	if (manager == NULL) {
+		return;
+	}
+
+	struct wlr_foreign_toplevel_v1 *toplevel, *tmp_toplevel;
+	wl_list_for_each_safe(toplevel, tmp_toplevel, &manager->toplevels, link) {
```

Comment:
I wasn't sure about that, we shouldn't auto-destroy things that the compositor created (the `wlr_foreign_toplevel_v1`)?